web3/eth.py

```diff
@@ -397,39 +467,39 @@ def filter(self, filter_params=None, filter_id=None):
                             "a valid filter object, or a filter_id as a string "
                             "or hex.")
 
-    def getFilterChanges(self, filter_id):
+    def getFilterChanges(self, filter_id: int) -> List[LogParams]:
         return self.web3.manager.request_blocking(
             "eth_getFilterChanges", [filter_id],
         )
 
-    def getFilterLogs(self, filter_id):
+    def getFilterLogs(self, filter_id: int) -> List[LogParams]:
         return self.web3.manager.request_blocking(
             "eth_getFilterLogs", [filter_id],
         )
 
-    def getLogs(self, filter_params):
+    def getLogs(self, filter_params: FilterParams) -> List[LogParams]:
         return self.web3.manager.request_blocking(
             "eth_getLogs", [filter_params],
         )
 
-    def submitHashrate(self, hashrate, node_id):
+    def submitHashrate(self, hashrate: int, node_id: Hash32) -> bool:
         return self.web3.manager.request_blocking(
             "eth_submitHashrate", [hashrate, node_id],
         )
 
-    def submitWork(self, nonce, pow_hash, mix_digest):
+    def submitWork(self, nonce: int, pow_hash: Hash32, mix_digest: Hash32) -> bool:
         return self.web3.manager.request_blocking(
             "eth_submitWork", [nonce, pow_hash, mix_digest],
         )
 
-    def uninstallFilter(self, filter_id):
+    def uninstallFilter(self, filter_id: int) -> bool:
         return self.web3.manager.request_blocking(
             "eth_uninstallFilter", [filter_id],
         )
 
-    def contract(self,
-                 address=None,
-                 **kwargs):
+    def contract(
+        self, address: Union[Address, ChecksumAddress, ENS]=None, **kwargs: Any
+    ) -> Any:
```

Comment:
This return type is particularly confusing, AFAIK it should be something like: `Union[Contract, ConciseContract, ContractCaller, Type[Union[Contract, ConciseContract, ContractCaller]]]` - however that breaks about a million things throughout the entire codebase, as does pretty much everything else I've tried. Also, b/c `ContractCaller` doesn't implement a factory, should it be allowed in here / `setContractFactory` below?

Reply:
You might be able to make this work with `typing.overload`.  Define one where `address`: None` and have that return the `Type[Contract]`, and another that has `address: Union[Address, ...]` and have it return `Contract`

Alternate approach would be to just *force* the type to be `Contract` and pretend the other types don't exist via either `cast` or `# type: ignore`.